src/i_midimusic.c

```diff
@@ -138,273 +123,82 @@ typedef struct
+        I_Printf(VB_ERROR, "SendShortMsg: %s", midiout->msg);
     }
 }
 
-// Writes a short MIDI message. Call this function from the MIDI thread only,
-// with exclusive access to shared resources.
-
-static void SendShortMsg(unsigned int delta_time, byte status, byte channel,
-                         byte param1, byte param2)
+static void SendChannelMsg(const midi_event_t *event, boolean use_param2)
 {
-    native_event_t native_event;
-    native_event.dwDeltaTime = delta_time;
-    native_event.dwStreamID = 0;
-    native_event.dwEvent =
-        MAKE_EVT(status | channel, param1, param2, MEVT_SHORTMSG);
-    WriteBuffer((byte *)&native_event, sizeof(native_event_t));
-}
-
-// Writes a short MIDI message (from an event). Call this function from the MIDI
-// thread only, with exclusive access to shared resources.
-
-static void SendChannelMsg(unsigned int delta_time, const midi_event_t *event,
-                           boolean use_param2)
-{
-    SendShortMsg(delta_time, event->event_type, event->data.channel.channel,
+    SendShortMsg(event->event_type, event->data.channel.channel,
                  event->data.channel.param1,
                  use_param2 ? event->data.channel.param2 : 0);
 }
 
-// Writes a long MIDI message (SysEx). Call this function from the MIDI thread
-// only, with exclusive access to shared resources.
-
-static void SendLongMsg(unsigned int delta_time, const byte *ptr,
-                        unsigned int length)
+static void SendLongMsg(const byte *ptr, unsigned int length)
 {
-    native_event_t native_event;
-    native_event.dwDeltaTime = delta_time;
-    native_event.dwStreamID = 0;
-    native_event.dwEvent = MAKE_EVT(length, 0, 0, MEVT_LONGMSG);
-    WriteBuffer((byte *)&native_event, sizeof(native_event_t));
-    WriteBuffer(ptr, length);
-    WriteBufferPad();
+    if (rtmidi_out_send_message(midiout, ptr, length) < 0)
+    {
+        I_Printf(VB_ERROR, "SendLongMsg: %s", midiout->msg);
+    }
 }
 
 // Writes an RPN message set to NULL (0x7F). Prevents accidental data entry.
-// Call this function from the MIDI thread only, with exclusive access to shared
-// resources.
 
-static void SendNullRPN(unsigned int delta_time, const midi_event_t *event)
+static void SendNullRPN(const midi_event_t *event)
 {
     const byte channel = event->data.channel.channel;
-    SendShortMsg(delta_time, MIDI_EVENT_CONTROLLER, channel,
-                 MIDI_CONTROLLER_RPN_LSB, MIDI_RPN_NULL);
-    SendShortMsg(0, MIDI_EVENT_CONTROLLER, channel, MIDI_CONTROLLER_RPN_MSB,
+    SendShortMsg(MIDI_EVENT_CONTROLLER, channel, MIDI_CONTROLLER_RPN_LSB,
+                 MIDI_RPN_NULL);
+    SendShortMsg(MIDI_EVENT_CONTROLLER, channel, MIDI_CONTROLLER_RPN_MSB,
                  MIDI_RPN_NULL);
 }
 
-// Writes a NOP message (ticks). Call this function from the MIDI thread only,
-// with exclusive access to shared resources.
+// static void SendNOPMsg(unsigned int delta_time)
+// {
+//     ;
+// }
```

Comment:
@ceski-1 I commented out `SendNOPMsg`, do we need this function?

Reply:
It was used to preserve timing for events, e.g. looping correctly at the end of a track. I'm not sure if it's needed with the new library (I haven't reviewed or tested the PR).

Reply:
Now we send messages directly and manage the time ourselves. I vaguely remember looping problems, do you have test files?

Reply:
I'm almost certain the following will break: emidi songs, complex songs when using vanilla midi complevel, songs with loop points. I won't be able to do a thorough review for a while, sorry.

Reply:
I tested some MIDI's from FF VII (XG, Yamaha S-YXG50) and Shadow Warrior (EMIDI, MS GS Wavetable Synth). It seems to work fine, but I might be missing some details.

It shouldn't be too bad, I tried to preserve the logic.